main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1715 @@
+          case False => return False // If the conjunction contains FALSE then whole conjunct is FALSE.
+          case x@Term.Cst(_) => cstTerms += x
+          case x@Term.Var(_) => varTerms += x
+          case And(csts0, vars0, rest0) =>
+            // We have found a nested conjunction. We can immediately add _its_ constants and variables.
+            cstTerms ++= csts0
+            varTerms ++= vars0
+            for (t0 <- rest0) {
+              // We then iterate through the nested sub-terms of the nested conjunction.
+              t0 match {
+                case True => // NOP - We do not have to include True in a conjunction.
+                case False => return False // If the sub-conjunction contains FALSE then the whole conjunct is FALSE.
+                case x@Term.Cst(_) => cstTerms += x
+                case x@Term.Var(_) => varTerms += x
+                case _ => restTerms += t0
+              }
+            }
+          case _ => restTerms += t // We found some other sub-term.
+        }
+      }
+
+      // We now have a set of constants, a set of variables, and a list of sub-terms.
+      // We optimize for the case where each of these is empty EXCEPT for one element.
+      (cstTerms.toList, varTerms.toList, restTerms.toList) match {
+        case (Nil, Nil, Nil) => Term.True // Everything is empty, so we return the neutral element, i.e. TRUE.
+        case (List(c), Nil, Nil) => c // A single constant.
+        case (Nil, List(x), Nil) => x // A single variable.
+        case (Nil, Nil, List(t)) => t // A single non-constant and non-variable sub-term.
+        case _ => And(cstTerms.toSet, varTerms.toSet, restTerms.toList)
+      }
+    }
+
+    /**
+      * Smart constructor for disjunction.
+      */
+    final def mkOr(ts: List[Term]): Term = {
+      // We refer to [[mkAnd]] since the structure is similar.
+
+      val varTerms = mutable.Set.empty[Term]
+      val nonVarTerms = mutable.ListBuffer.empty[Term]
+      for (t <- ts) {
+        t match {
+          case True => return True
+          case False => // nop
+          case x@Term.Var(_) => varTerms += x
+          case Or(ts0) =>
+            for (t0 <- ts0) {
+              t0 match {
+                case True => return True
+                case False => // nop
+                case x@Term.Var(_) => varTerms += x
+                case _ => nonVarTerms += t0
+              }
+            }
+          case _ => nonVarTerms += t
+        }
+      }
+
+      varTerms.toList ++ nonVarTerms.toList match {
+        case Nil => False
+        case x :: Nil => x
+        case xs => Or(xs)
+      }
+    }
+
+    /**
+      * Returns the Xor of `x` and `y`. Implemented by desugaring.
+      */
+    final def mkXor(x: Term, y: Term): Term = mkOr(mkAnd(x, mkNot(y)), mkAnd(mkNot(x), y))
+
+  }
+
+  /**
+    * Companion object of [[BoolSubstitution]].
+    */
+  object BoolSubstitution {
```

Comment:
Private? Or maybe move to its own file

Reply:
It cannot be private, it is used in EffUnification2.

Reply:
I am keeping everything in one file for now-- because we may want to duplicate for set effects. Later I plan to split into multiple files. #7363